// @progress/kendo-angular-grid/esm2020/grid.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { GridSpacerComponent } from './rendering/common/spacer.component';
import { GridMarqueeDirective } from './selection/marquee.directive';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { GridComponent } from './grid.component';
import { ListComponent } from './rendering/list.component';
import { DataBindingDirective } from './databinding.directive';
import { SelectionDirective } from './selection/selection.directive';
import { ExpandDetailsDirective } from './rendering/details-expand.directive';
import { ExpandGroupDirective } from './rendering/groups-expand.directive';
import { LocalizedMessagesDirective } from './localization/localized-messages.directive';
import { CustomMessagesComponent } from './localization/custom-messages.component';
import { RowFilterModule } from "./filtering/cell/row-filtering.module";
import { PagerModule } from "./pager/pager.module";
import { GroupModule } from "./grouping/group.module";
import { HeaderModule } from "./rendering/header/header.module";
import { BodyModule } from "./rendering/body.module";
import { FooterModule } from "./rendering/footer/footer.module";
import { SharedModule } from './shared.module';
import { ToolbarTemplateDirective } from "./rendering/toolbar/toolbar-template.directive";
import { ToolbarComponent } from "./rendering/toolbar/toolbar.component";
import { StatusBarTemplateDirective } from './aggregates/status-bar-template.directive';
import { StatusBarComponent } from './aggregates/status-bar.component';
import { ResizeSensorModule, WatermarkModule } from '@progress/kendo-angular-common';
import { TemplateEditingDirective } from './editing-directives/template-editing.directive';
import { ReactiveEditingDirective } from './editing-directives/reactive-editing.directive';
import { InCellEditingDirective } from './editing-directives/in-cell-editing.directive';
import { GroupBindingDirective } from './grouping/group-scroll-binding.directive';
import { FilterMenuModule } from "./filtering/menu/filter-menu.module";
import { ColumnMenuModule } from './column-menu/column-menu.module';
import { GridToolbarFocusableDirective } from './rendering/toolbar/toolbar-focusable.directive';
import { DragAndDropModule } from '@progress/kendo-angular-utils';
import { GridClipboardDirective } from './common/clipboard.directive';
import * as i0 from "@angular/core";
import * as i1 from "./grouping/group-header-template.directive";
import * as i2 from "./grouping/group-header-column-template.directive";
import * as i3 from "./grouping/group-footer-template.directive";
import * as i4 from "./columns/column.component";
import * as i5 from "./columns/span-column.component";
import * as i6 from "./columns/column-group.component";
import * as i7 from "./rendering/footer/footer-template.directive";
import * as i8 from "./rendering/details/detail-template.directive";
import * as i9 from "./navigation/focusable.directive";
import * as i10 from "./rendering/grid-table.directive";
import * as i11 from "./columns/command-column.component";
import * as i12 from "./columns/checkbox-column.component";
import * as i13 from "./selection/selection-checkbox.directive";
import * as i14 from "./rendering/cell-template.directive";
import * as i15 from "./rendering/no-records-template.directive";
import * as i16 from "./editing/edit-template.directive";
import * as i17 from "./editing/edit-command.directive";
import * as i18 from "./editing/cancel-command.directive";
import * as i19 from "./editing/save-command.directive";
import * as i20 from "./editing/remove-command.directive";
import * as i21 from "./editing/add-command.directive";
import * as i22 from "./rendering/cell-loading.template.directive";
import * as i23 from "./rendering/loading-template.directive";
import * as i24 from "./columns/reorder-column.component";
import * as i25 from "./rendering/header/header-template.directive";
import * as i26 from "./selection/selectall-checkbox.directive";
import * as i27 from "./pager/pager.component";
import * as i28 from "./pager/pager-prev-buttons.component";
import * as i29 from "./pager/pager-next-buttons.component";
import * as i30 from "./pager/pager-numeric-buttons.component";
import * as i31 from "./pager/pager-input.component";
import * as i32 from "./pager/pager-info.component";
import * as i33 from "./pager/pager-page-sizes.component";
import * as i34 from "./pager/pager-template.directive";
import * as i35 from "./pager/pager-dropdown.directive";
import * as i36 from "./pager/pager-input.directive";
import * as i37 from "./filtering/filter-row.component";
import * as i38 from "./filtering/cell/filter-cell.component";
import * as i39 from "./filtering/cell/filter-cell-template.directive";
import * as i40 from "./filtering/cell/filter-cell-operators.component";
import * as i41 from "./filtering/cell/string-filter-cell.component";
import * as i42 from "./filtering/cell/numeric-filter-cell.component";
import * as i43 from "./filtering/cell/autocomplete-filter-cell.component";
import * as i44 from "./filtering/cell/boolean-filter-cell.component";
import * as i45 from "./filtering/cell/date-filter-cell.component";
import * as i46 from "./filtering/operators/contains-filter-operator.component";
import * as i47 from "./filtering/operators/not-contains-filter-operator.component";
import * as i48 from "./filtering/operators/ends-with-filter-operator.component";
import * as i49 from "./filtering/operators/eq-filter-operator.component";
import * as i50 from "./filtering/operators/is-empty-filter-operator.component";
import * as i51 from "./filtering/operators/is-not-empty-filter-operator.component";
import * as i52 from "./filtering/operators/is-not-null-filter-operator.component";
import * as i53 from "./filtering/operators/isnull-filter-operator.component";
import * as i54 from "./filtering/operators/neq-filter-operator.component";
import * as i55 from "./filtering/operators/starts-with-filter-operator.component";
import * as i56 from "./filtering/operators/gt-filter-operator.component";
import * as i57 from "./filtering/operators/gte-filter-operator.component";
import * as i58 from "./filtering/operators/lt-filter-operator.component";
import * as i59 from "./filtering/operators/lte-filter-operator.component";
import * as i60 from "./filtering/operators/after-filter-operator.component";
import * as i61 from "./filtering/operators/after-eq-filter-operator.component";
import * as i62 from "./filtering/operators/before-eq-filter-operator.component";
import * as i63 from "./filtering/operators/before-filter-operator.component";
import * as i64 from "./filtering/menu/string-filter-menu.component";
import * as i65 from "./filtering/menu/filter-menu-template.directive";
import * as i66 from "./filtering/menu/numeric-filter-menu.component";
import * as i67 from "./filtering/menu/date-filter-menu.component";
import * as i68 from "./filtering/menu/boolean-filter-menu.component";
import * as i69 from "./filtering/menu/filter-menu-dropdownlist.directive";
import * as i70 from "./filtering/menu/filter-radio-button.directive";
import * as i71 from "./column-menu/column-chooser.component";
import * as i72 from "./column-menu/column-menu-filter.component";
import * as i73 from "./column-menu/column-menu-item.component";
import * as i74 from "./column-menu/column-menu-item-content-template.directive";
import * as i75 from "./column-menu/column-menu-sort.component";
import * as i76 from "./column-menu/column-menu-lock.component";
import * as i77 from "./column-menu/column-menu-stick.component";
import * as i78 from "./column-menu/column-menu-position.component";
import * as i79 from "./column-menu/column-menu-chooser.component";
import * as i80 from "./column-menu/column-menu-template.directive";
import * as i81 from "./column-menu/column-menu-container.component";
import * as i82 from "./column-menu/column-menu-item.directive";
import * as i83 from "./column-menu/column-menu.component";
import * as i84 from "./column-menu/column-menu-autosize.component";
import * as i85 from "./column-menu/column-menu-autosize-all.component";
const exportedModules = [
    GridComponent,
    ToolbarTemplateDirective,
    ToolbarComponent,
    GridSpacerComponent,
    StatusBarTemplateDirective,
    DataBindingDirective,
    SelectionDirective,
    CustomMessagesComponent,
    GroupBindingDirective,
    TemplateEditingDirective,
    ReactiveEditingDirective,
    InCellEditingDirective,
    ExpandDetailsDirective,
    ExpandGroupDirective,
    GridToolbarFocusableDirective,
    ...GroupModule.exports(),
    ...SharedModule.exports(),
    ...BodyModule.exports(),
    ...HeaderModule.exports(),
    ...FooterModule.exports(),
    ...PagerModule.exports(),
    ...RowFilterModule.exports(),
    ...FilterMenuModule.exports(),
    ...ColumnMenuModule.exports(),
    GridClipboardDirective
];
const declarations = [
    GridComponent,
    ListComponent,
    ToolbarComponent,
    LocalizedMessagesDirective,
    CustomMessagesComponent,
    DataBindingDirective,
    ToolbarTemplateDirective,
    SelectionDirective,
    TemplateEditingDirective,
    ReactiveEditingDirective,
    InCellEditingDirective,
    ExpandDetailsDirective,
    ExpandGroupDirective,
    GroupBindingDirective,
    GridMarqueeDirective,
    GridSpacerComponent,
    GridToolbarFocusableDirective,
    StatusBarComponent,
    StatusBarTemplateDirective,
    GridClipboardDirective
];
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Grid component.
 *
 * @example
 *
 * ```ts
 * import { NgModule } from '@angular/core';
 * import { GridModule } from '@progress/kendo-angular-grid';
 * import { BrowserModule } from "@angular/platform-browser";
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, GridModule],
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * ```
 */
export class GridModule {
}
GridModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: GridModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
GridModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: GridModule, declarations: [GridComponent,
        ListComponent,
        ToolbarComponent,
        LocalizedMessagesDirective,
        CustomMessagesComponent,
        DataBindingDirective,
        ToolbarTemplateDirective,
        SelectionDirective,
        TemplateEditingDirective,
        ReactiveEditingDirective,
        InCellEditingDirective,
        ExpandDetailsDirective,
        ExpandGroupDirective,
        GroupBindingDirective,
        GridMarqueeDirective,
        GridSpacerComponent,
        GridToolbarFocusableDirective,
        StatusBarComponent,
        StatusBarTemplateDirective,
        GridClipboardDirective], imports: [CommonModule,
        GroupModule,
        SharedModule,
        BodyModule,
        HeaderModule,
        FooterModule,
        PagerModule,
        RowFilterModule,
        FilterMenuModule,
        ResizeSensorModule,
        ColumnMenuModule,
        DragAndDropModule,
        WatermarkModule], exports: [GridComponent,
        ToolbarTemplateDirective,
        ToolbarComponent,
        GridSpacerComponent,
        StatusBarTemplateDirective,
        DataBindingDirective,
        SelectionDirective,
        CustomMessagesComponent,
        GroupBindingDirective,
        TemplateEditingDirective,
        ReactiveEditingDirective,
        InCellEditingDirective,
        ExpandDetailsDirective,
        ExpandGroupDirective,
        GridToolbarFocusableDirective, i1.GroupHeaderTemplateDirective, i2.GroupHeaderColumnTemplateDirective, i3.GroupFooterTemplateDirective, i4.ColumnComponent, i5.SpanColumnComponent, i6.ColumnGroupComponent, i7.FooterTemplateDirective, i8.DetailTemplateDirective, i9.FocusableDirective, i10.GridTableDirective, i11.CommandColumnComponent, i12.CheckboxColumnComponent, i13.SelectionCheckboxDirective, i14.CellTemplateDirective, i15.NoRecordsTemplateDirective, i16.EditTemplateDirective, i17.EditCommandDirective, i18.CancelCommandDirective, i19.SaveCommandDirective, i20.RemoveCommandDirective, i21.AddCommandDirective, i22.CellLoadingTemplateDirective, i23.LoadingTemplateDirective, i24.RowReorderColumnComponent, i25.HeaderTemplateDirective, i26.SelectAllCheckboxDirective, i27.PagerComponent, i28.PagerPrevButtonsComponent, i29.PagerNextButtonsComponent, i30.PagerNumericButtonsComponent, i31.PagerInputComponent, i32.PagerInfoComponent, i33.PagerPageSizesComponent, i34.PagerTemplateDirective, i35.PagerDropDownListDirective, i36.PagerInputDirective, i37.FilterRowComponent, i38.FilterCellComponent, i39.FilterCellTemplateDirective, i40.FilterCellOperatorsComponent, i41.StringFilterCellComponent, i42.NumericFilterCellComponent, i43.AutoCompleteFilterCellComponent, i44.BooleanFilterCellComponent, i45.DateFilterCellComponent, i40.FilterCellOperatorsComponent, i46.ContainsFilterOperatorComponent, i47.DoesNotContainFilterOperatorComponent, i48.EndsWithFilterOperatorComponent, i49.EqualFilterOperatorComponent, i50.IsEmptyFilterOperatorComponent, i51.IsNotEmptyFilterOperatorComponent, i52.IsNotNullFilterOperatorComponent, i53.IsNullFilterOperatorComponent, i54.NotEqualFilterOperatorComponent, i55.StartsWithFilterOperatorComponent, i56.GreaterFilterOperatorComponent, i57.GreaterOrEqualToFilterOperatorComponent, i58.LessFilterOperatorComponent, i59.LessOrEqualToFilterOperatorComponent, i60.AfterFilterOperatorComponent, i61.AfterEqFilterOperatorComponent, i62.BeforeEqFilterOperatorComponent, i63.BeforeFilterOperatorComponent, i64.StringFilterMenuComponent, i65.FilterMenuTemplateDirective, i66.NumericFilterMenuComponent, i67.DateFilterMenuComponent, i68.BooleanFilterMenuComponent, i69.FilterMenuDropDownListDirective, i70.BooleanFilterRadioButtonDirective, i40.FilterCellOperatorsComponent, i46.ContainsFilterOperatorComponent, i47.DoesNotContainFilterOperatorComponent, i48.EndsWithFilterOperatorComponent, i49.EqualFilterOperatorComponent, i50.IsEmptyFilterOperatorComponent, i51.IsNotEmptyFilterOperatorComponent, i52.IsNotNullFilterOperatorComponent, i53.IsNullFilterOperatorComponent, i54.NotEqualFilterOperatorComponent, i55.StartsWithFilterOperatorComponent, i56.GreaterFilterOperatorComponent, i57.GreaterOrEqualToFilterOperatorComponent, i58.LessFilterOperatorComponent, i59.LessOrEqualToFilterOperatorComponent, i60.AfterFilterOperatorComponent, i61.AfterEqFilterOperatorComponent, i62.BeforeEqFilterOperatorComponent, i63.BeforeFilterOperatorComponent, i71.ColumnChooserComponent, i72.ColumnMenuFilterComponent, i73.ColumnMenuItemComponent, i74.ColumnMenuItemContentTemplateDirective, i75.ColumnMenuSortComponent, i76.ColumnMenuLockComponent, i77.ColumnMenuStickComponent, i78.ColumnMenuPositionComponent, i79.ColumnMenuChooserComponent, i80.ColumnMenuTemplateDirective, i81.ColumnMenuContainerComponent, i82.ColumnMenuItemDirective, i83.ColumnMenuComponent, i84.ColumnMenuAutoSizeColumnComponent, i85.ColumnMenuAutoSizeAllColumnsComponent, GridClipboardDirective] });
GridModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: GridModule, imports: [[
            CommonModule,
            GroupModule,
            SharedModule,
            BodyModule,
            HeaderModule,
            FooterModule,
            PagerModule,
            RowFilterModule,
            FilterMenuModule,
            ResizeSensorModule,
            ColumnMenuModule,
            DragAndDropModule,
            WatermarkModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: GridModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [declarations],
                    exports: [exportedModules],
                    imports: [
                        CommonModule,
                        GroupModule,
                        SharedModule,
                        BodyModule,
                        HeaderModule,
                        FooterModule,
                        PagerModule,
                        RowFilterModule,
                        FilterMenuModule,
                        ResizeSensorModule,
                        ColumnMenuModule,
                        DragAndDropModule,
                        WatermarkModule
                    ]
                }]
        }] });
