// @progress/kendo-angular-inputs/esm2020/textarea.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { TextAreaComponent } from "./textarea/textarea.component";
import { AdornmentsModule, EventsModule } from "@progress/kendo-angular-common";
import { TextAreaPrefixComponent } from "./textarea/textarea-prefix.component";
import { TextAreaSuffixComponent } from './textarea/textarea-suffix.component';
import { SharedModule, SHARED_DIRECTIVES } from './shared.module';
import { SharedEventsModule } from "./shared-events.module";
import * as i0 from "@angular/core";
import * as i1 from "./shared/input-separator.component";
import * as i2 from "./shared/textarea.directive";
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the TextArea component.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the TextArea module
 * import { TextAreaModule } from '@progress/kendo-angular-inputs';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, TextAreaModule], // import TextArea module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class TextAreaModule {
}
TextAreaModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: TextAreaModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TextAreaModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: TextAreaModule, declarations: [TextAreaComponent,
        TextAreaPrefixComponent,
        TextAreaSuffixComponent], imports: [CommonModule, EventsModule, SharedModule, AdornmentsModule, SharedEventsModule], exports: [TextAreaComponent,
        EventsModule,
        TextAreaPrefixComponent,
        TextAreaSuffixComponent,
        AdornmentsModule, i1.InputSeparatorComponent, i2.TextAreaDirective] });
TextAreaModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: TextAreaModule, imports: [[CommonModule, EventsModule, SharedModule, AdornmentsModule, SharedEventsModule], EventsModule,
        AdornmentsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: TextAreaModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        TextAreaComponent,
                        TextAreaPrefixComponent,
                        TextAreaSuffixComponent
                    ],
                    exports: [
                        TextAreaComponent,
                        EventsModule,
                        TextAreaPrefixComponent,
                        TextAreaSuffixComponent,
                        AdornmentsModule,
                        SHARED_DIRECTIVES
                    ],
                    imports: [CommonModule, EventsModule, SharedModule, AdornmentsModule, SharedEventsModule]
                }]
        }] });
