// @progress/kendo-angular-grid/esm2020/aggregates/status-bar.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, HostBinding, Input } from '@angular/core';
import { StatusBarTemplateDirective } from './status-bar-template.directive';
import { CellSelectionAggregateService } from './selection-aggregate.service';
import * as i0 from "@angular/core";
import * as i1 from "./selection-aggregate.service";
import * as i2 from "@angular/common";
/**
 * @hidden
 */
export class StatusBarComponent {
    constructor(aggregateService) {
        this.aggregateService = aggregateService;
        this.hostClasses = true;
    }
    get isStatusBarTemplate() {
        return this.statusBarTemplate ? '' : 'none';
    }
    ngOnInit() {
        this.aggregateService.init();
    }
    get aggregates() {
        return this.aggregateService.aggregates;
    }
}
StatusBarComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: StatusBarComponent, deps: [{ token: i1.CellSelectionAggregateService }], target: i0.ɵɵFactoryTarget.Component });
StatusBarComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.12", type: StatusBarComponent, selector: "kendo-grid-status-bar", inputs: { statusBarTemplate: "statusBarTemplate" }, host: { properties: { "class.k-selection-aggregates": "this.hostClasses", "class.k-grid-selection-aggregates": "this.hostClasses", "style.display": "this.isStatusBarTemplate" } }, ngImport: i0, template: `
        <ng-template
            *ngIf="statusBarTemplate"
            [ngTemplateOutlet]="statusBarTemplate?.templateRef"
            [ngTemplateOutletContext]="{ aggregates: aggregates }">
        </ng-template>
    `, isInline: true, directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i2.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: StatusBarComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'kendo-grid-status-bar',
                    template: `
        <ng-template
            *ngIf="statusBarTemplate"
            [ngTemplateOutlet]="statusBarTemplate?.templateRef"
            [ngTemplateOutletContext]="{ aggregates: aggregates }">
        </ng-template>
    `
                }]
        }], ctorParameters: function () { return [{ type: i1.CellSelectionAggregateService }]; }, propDecorators: { hostClasses: [{
                type: HostBinding,
                args: ['class.k-selection-aggregates']
            }, {
                type: HostBinding,
                args: ['class.k-grid-selection-aggregates']
            }], isStatusBarTemplate: [{
                type: HostBinding,
                args: ['style.display']
            }], statusBarTemplate: [{
                type: Input
            }] } });
